File: drone-3d-printing-design/firmware/python/drone_firmware.py
"""
Python Drone Firmware — MicroPython / CircuitPython compatible

Targets:
  • Raspberry Pi Pico (RP2040) with MicroPython
  • Adafruit Feather M4 with CircuitPython
  • Any Linux SBC (full CPython) for SITL

Usage (MicroPython on Pico):
  Copy this file to the Pico as main.py.
  Wire MPU-6050 to I2C0 (SDA=GP0, SCL=GP1).
  Wire 4 ESCs to GP2, GP3, GP4, GP5.

The module is structured to work both as a standalone firmware
(if __name__ == "__main__") and as an importable library for testing.
"""

# ---------------------------------------------------------------------------
# MicroPython / CircuitPython compatibility shim
# Try real hardware modules; fall back to simulation stubs when not available.
# ---------------------------------------------------------------------------

try:
    import machine          # type: ignore  # MicroPython hardware
    import time
    _ON_HARDWARE = True
except ImportError:
    import time
    _ON_HARDWARE = False


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

PWM_FREQ_HZ  = 50    # standard servo/ESC PWM frequency
PWM_MIN_US   = 1000  # motor stopped
PWM_MAX_US   = 2000  # full throttle
LOOP_RATE_HZ = 200   # control loop frequency


# ---------------------------------------------------------------------------
# Hardware abstraction layer (HAL)
# ---------------------------------------------------------------------------

class PWMOutput:
    """
    Thin wrapper around machine.PWM (MicroPython) or a simulation stub.
    """

    def __init__(self, pin_number: int) -> None:
        self.pin_number = pin_number
        self._us = PWM_MIN_US

        if _ON_HARDWARE:
            pin = machine.Pin(pin_number)
            self._pwm = machine.PWM(pin)
            self._pwm.freq(PWM_FREQ_HZ)
            self._set_us(PWM_MIN_US)

    def _set_us(self, microseconds: int) -> None:
        """Convert µs pulse width to PWM duty cycle."""
        self._us = microseconds
        if _ON_HARDWARE:
            # MicroPython: duty_ns or duty_u16
            period_ns = 1_000_000_000 // PWM_FREQ_HZ   # 20 ms = 20,000,000 ns
            duty_ns   = microseconds * 1000
            self._pwm.duty_ns(min(duty_ns, period_ns - 1))

    def write(self, microseconds: int) -> None:
        us = max(PWM_MIN_US, min(PWM_MAX_US, microseconds))
        self._set_us(us)

    @property
    def current_us(self) -> int:
        return self._us

    def deinit(self) -> None:
        if _ON_HARDWARE:
            self._pwm.deinit()


class I2CDevice:
    """Stub wrapper for I2C peripheral (MPU-6050, BMP388, etc.)."""

    def __init__(self, address: int, i2c_id: int = 0) -> None:
        self.address = address
        self._i2c = None
        if _ON_HARDWARE:
            sda = machine.Pin(0)
            scl = machine.Pin(1)
            self._i2c = machine.I2C(i2c_id, sda=sda, scl=scl, freq=400_000)

    def read_register(self, reg: int, length: int = 1) -> bytes:
        if _ON_HARDWARE and self._i2c:
            return self._i2c.readfrom_mem(self.address, reg, length)
        # Simulation: return zeroed bytes
        return bytes(length)

    def write_register(self, reg: int, data: bytes) -> None:
        if _ON_HARDWARE and self._i2c:
            self._i2c.writeto_mem(self.address, reg, data)


# ---------------------------------------------------------------------------
# MPU-6050 driver (simplified)
# ---------------------------------------------------------------------------

MPU6050_ADDR      = 0x68
MPU6050_PWR_MGMT  = 0x6B
MPU6050_ACCEL_OUT = 0x3B
MPU6050_GYRO_OUT  = 0x43
ACCEL_SCALE       = 16384.0   # ±2g → LSB/g
GYRO_SCALE        = 131.0     # ±250 °/s → LSB/(°/s)


class MPU6050:
    """Minimal MPU-6050 driver for MicroPython."""

    def __init__(self) -> None:
        self._dev = I2CDevice(MPU6050_ADDR)
        # Wake up device
        self._dev.write_register(MPU6050_PWR_MGMT, b'\x00')

    def _to_signed(self, high: int, low: int) -> int:
        val = (high << 8) | low
        return val - 65536 if val > 32767 else val

    def read_accel_gyro(self):
        """
        Returns (ax, ay, az, gx, gy, gz).
        Accelerometer in m/s², gyro in deg/s.
        """
        raw = self._dev.read_register(MPU6050_ACCEL_OUT, 14)
        if len(raw) < 14:
            return (0.0,) * 6

        ax = self._to_signed(raw[0],  raw[1])  / ACCEL_SCALE * 9.81
        ay = self._to_signed(raw[2],  raw[3])  / ACCEL_SCALE * 9.81
        az = self._to_signed(raw[4],  raw[5])  / ACCEL_SCALE * 9.81
        gx = self._to_signed(raw[8],  raw[9])  / GYRO_SCALE
        gy = self._to_signed(raw[10], raw[11]) / GYRO_SCALE
        gz = self._to_signed(raw[12], raw[13]) / GYRO_SCALE

        return ax, ay, az, gx, gy, gz


# ---------------------------------------------------------------------------
# Minimal PID
# ---------------------------------------------------------------------------

class PID:
    def __init__(self, kp: float, ki: float, kd: float,
                 limit: float = 400.0) -> None:
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.limit = limit
        self._integral  = 0.0
        self._prev_error = 0.0

    def compute(self, setpoint: float, measured: float, dt: float) -> float:
        err = setpoint - measured
        self._integral = max(-self.limit,
                             min(self.limit, self._integral + err * dt))
        d_term = (err - self._prev_error) / dt if dt > 0 else 0.0
        self._prev_error = err
        out = self.kp * err + self.ki * self._integral + self.kd * d_term
        return max(-self.limit, min(self.limit, out))

    def reset(self) -> None:
        self._integral   = 0.0
        self._prev_error = 0.0


# ---------------------------------------------------------------------------
# Complementary filter
# ---------------------------------------------------------------------------

import math as _math


class CompFilter:
    def __init__(self, alpha: float = 0.98) -> None:
        self.alpha = alpha
        self.roll  = 0.0
        self.pitch = 0.0

    def update(self, ax, ay, az, gx, gy, dt):
        self.roll  = self.alpha * (self.roll  + gx * dt)
        self.pitch = self.alpha * (self.pitch + gy * dt)
        if az != 0:
            ar = _math.degrees(_math.atan2(ay, az))
            ap = _math.degrees(_math.atan2(-ax, _math.sqrt(ay**2 + az**2)))
            self.roll  += (1 - self.alpha) * ar
            self.pitch += (1 - self.alpha) * ap
        return self.roll, self.pitch


# ---------------------------------------------------------------------------
# Firmware main loop
# ---------------------------------------------------------------------------

def run_firmware(loop_count: int = 100) -> None:
    """
    Main firmware loop.

    Parameters
    ----------
    loop_count : int
        Number of iterations to run.  Pass -1 for infinite (hardware use).
    """
    # Initialise hardware
    motors = [PWMOutput(2), PWMOutput(3), PWMOutput(4), PWMOutput(5)]
    imu    = MPU6050()
    filt   = CompFilter(alpha=0.98)
    pids   = {
        "roll":  PID(4.5, 0.04, 0.40),
        "pitch": PID(4.5, 0.04, 0.40),
        "yaw":   PID(6.0, 0.10, 0.20),
    }

    dt = 1.0 / LOOP_RATE_HZ
    armed = False
    throttle_pct = 0.0

    iteration = 0
    while loop_count < 0 or iteration < loop_count:
        t_start = time.time()

        # --- Sensor read ---
        ax, ay, az, gx, gy, gz = imu.read_accel_gyro()

        # --- Attitude estimate ---
        roll, pitch = filt.update(ax, ay, az, gx, gy, dt)

        # --- PID (setpoints = 0 → auto-level) ---
        roll_cmd  = pids["roll"].compute(0.0, roll,  dt)
        pitch_cmd = pids["pitch"].compute(0.0, pitch, dt)
        yaw_cmd   = pids["yaw"].compute(0.0, gz,     dt)

        # --- Motor mix (quad-X) ---
        norm = 500.0
        raw = [
            throttle_pct + pitch_cmd / norm + roll_cmd / norm - yaw_cmd / norm,
            throttle_pct + pitch_cmd / norm - roll_cmd / norm + yaw_cmd / norm,
            throttle_pct - pitch_cmd / norm - roll_cmd / norm - yaw_cmd / norm,
            throttle_pct - pitch_cmd / norm + roll_cmd / norm + yaw_cmd / norm,
        ]

        if armed:
            for motor, v in zip(motors, raw):
                motor.write(int(PWM_MIN_US + max(0.0, min(1.0, v))
                                * (PWM_MAX_US - PWM_MIN_US)))
        else:
            for motor in motors:
                motor.write(PWM_MIN_US)

        iteration += 1

        # Rate limiting (simulation only — on hardware use timer interrupts)
        elapsed = time.time() - t_start
        if elapsed < dt:
            time.sleep(dt - elapsed)

    # Cleanup
    for motor in motors:
        motor.write(PWM_MIN_US)
        motor.deinit()


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    print("Python Drone Firmware — simulation mode")
    print(f"Running {LOOP_RATE_HZ} Hz control loop for 50 iterations...\n")
    run_firmware(loop_count=50)
    print("Done.")

File: drone-3d-printing-design/firmware/tinygo/drone_firmware.go
// Drone Firmware — TinyGo
//
// Targets:
//   • Arduino Nano / Uno  (ATmega328P)   tinygo flash -target arduino
//   • Raspberry Pi Pico   (RP2040)       tinygo flash -target pico
//   • STM32F103 Blue Pill               tinygo flash -target stm32f103xx
//   • BBC micro:bit v2   (nRF52833)      tinygo flash -target microbit-v2
//
// Build:
//   tinygo build -o drone.uf2 -target pico .
//   tinygo flash  -target pico .
//
// Simulate on desktop (no hardware needed):
//   tinygo run .
//
// Dependencies:
//   tinygo.org/x/drivers — hardware drivers (I2C, SPI, UART)
//   Install:  go get tinygo.org/x/drivers

package main

import (
	"machine"
	"math"
	"time"
)

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const (
	pwmFreqHz  = 50    // ESC / servo PWM frequency
	pwmMinUs   = 1000  // motor stopped (µs)
	pwmMaxUs   = 2000  // full throttle (µs)
	loopRateHz = 200   // control loop frequency
	loopDtMs   = 1000 / loopRateHz
)

// ---------------------------------------------------------------------------
// PID controller
// ---------------------------------------------------------------------------

// PID holds the state for a single PID axis.
type PID struct {
	Kp, Ki, Kd float64
	Limit       float64 // anti-windup + output clamp
	integral    float64
	prevError   float64
}

// Compute returns the PID output for the given setpoint / measured pair.
func (p *PID) Compute(setpoint, measured, dt float64) float64 {
	if dt <= 0 {
		return 0
	}
	err := setpoint - measured

	p.integral += err * dt
	if p.integral > p.Limit {
		p.integral = p.Limit
	} else if p.integral < -p.Limit {
		p.integral = -p.Limit
	}

	dTerm := (err - p.prevError) / dt
	p.prevError = err

	out := p.Kp*err + p.Ki*p.integral + p.Kd*dTerm
	if out > p.Limit {
		return p.Limit
	} else if out < -p.Limit {
		return -p.Limit
	}
	return out
}

// Reset clears integrator and derivative state (call on disarm).
func (p *PID) Reset() {
	p.integral  = 0
	p.prevError = 0
}

// ---------------------------------------------------------------------------
// Complementary filter
// ---------------------------------------------------------------------------

// CompFilter fuses gyro and accelerometer to estimate roll / pitch.
type CompFilter struct {
	Alpha float64
	Roll  float64
	Pitch float64
}

// Update performs one filter iteration; returns (roll, pitch) in degrees.
func (f *CompFilter) Update(ax, ay, az, gx, gy, dt float64) (float64, float64) {
	f.Roll  += gx * dt
	f.Pitch += gy * dt

	if az != 0 {
		accelRoll  := math.Atan2(ay, az) * 180 / math.Pi
		accelPitch := math.Atan2(-ax, math.Sqrt(ay*ay+az*az)) * 180 / math.Pi
		f.Roll  = f.Alpha*f.Roll  + (1-f.Alpha)*accelRoll
		f.Pitch = f.Alpha*f.Pitch + (1-f.Alpha)*accelPitch
	}

	return f.Roll, f.Pitch
}

// ---------------------------------------------------------------------------
// Motor output helpers
// ---------------------------------------------------------------------------

// clamp01 constrains v to [0, 1].
func clamp01(v float64) float64 {
	if v < 0 {
		return 0
	}
	if v > 1 {
		return 1
	}
	return v
}

// throttleToUs converts a 0–1 throttle value to ESC µs (1000–2000).
func throttleToUs(v float64) uint32 {
	return uint32(pwmMinUs + clamp01(v)*float64(pwmMaxUs-pwmMinUs))
}

// mixQuadX returns motor PWM values [FL, FR, RR, RL] for a quad-X frame.
func mixQuadX(throttle, pitch, roll, yaw float64) [4]uint32 {
	norm := 500.0
	raw := [4]float64{
		throttle + pitch/norm + roll/norm - yaw/norm,
		throttle + pitch/norm - roll/norm + yaw/norm,
		throttle - pitch/norm - roll/norm - yaw/norm,
		throttle - pitch/norm + roll/norm + yaw/norm,
	}
	var out [4]uint32
	for i, v := range raw {
		out[i] = throttleToUs(v)
	}
	return out
}

// ---------------------------------------------------------------------------
// PWM output (hardware abstraction)
// ---------------------------------------------------------------------------

// setPWM writes a PWM pulse width to a machine.Pin.
// On hardware this uses the machine.PWM API.
// On desktop (tinygo run) it prints to stdout for inspection.
func setPWM(pin machine.Pin, us uint32) {
	// Real implementation — uncomment on actual hardware:
	//
	//   pwm := machine.PWM0   // choose correct PWM peripheral for the pin
	//   pwm.Configure(machine.PWMConfig{Period: 20_000_000})  // 50 Hz
	//   ch, _ := pwm.Channel(pin)
	//   pwm.Set(ch, pwm.Top()*us/20_000)
	//
	// Simulation stub:
	_ = pin
	_ = us
}

// ---------------------------------------------------------------------------
// Simulated IMU (replace with tinygo.org/x/drivers/mpu6050 on hardware)
// ---------------------------------------------------------------------------

// readIMU returns (ax, ay, az m/s², gx, gy, gz °/s).
// Simulation returns near-level values; replace with real driver call.
func readIMU() (ax, ay, az, gx, gy, gz float64) {
	// On hardware (Pico + MPU-6050 via I2C):
	//   sensor := mpu6050.New(machine.I2C0)
	//   sensor.Configure()
	//   data, _ := sensor.ReadAcceleration()
	//   ...
	return 0.0, 0.1, -9.81, 0.5, 0.2, 0.0
}

// ---------------------------------------------------------------------------
// RC input (simulated — replace with PPM/SBUS parser on hardware)
// ---------------------------------------------------------------------------

// RCInput holds normalised stick values.
type RCInput struct {
	Throttle float64 // 0–1
	Roll     float64 // -1 to +1
	Pitch    float64 // -1 to +1
	Yaw      float64 // -1 to +1
	Armed    bool
}

// readRC returns the latest RC input.
// Simulation returns a gentle hover command; replace with UART/PPM ISR.
func readRC() RCInput {
	return RCInput{
		Throttle: 0.45,
		Roll:     0.1,
		Pitch:    0.0,
		Yaw:      0.0,
		Armed:    true,
	}
}

// ---------------------------------------------------------------------------
// Motor pins (adjust for your wiring)
// ---------------------------------------------------------------------------

var motorPins = [4]machine.Pin{
	machine.GP2, // Front-Left
	machine.GP3, // Front-Right
	machine.GP4, // Rear-Right
	machine.GP5, // Rear-Left
}

// ---------------------------------------------------------------------------
// Main firmware loop
// ---------------------------------------------------------------------------

func main() {
	// Configure motor pins as output
	for _, pin := range motorPins {
		pin.Configure(machine.PinConfig{Mode: machine.PinOutput})
	}

	// Disarm pulse to all ESCs
	for _, pin := range motorPins {
		setPWM(pin, pwmMinUs)
	}
	time.Sleep(2 * time.Second) // ESC arming delay

	// Sensors & controllers
	filter := CompFilter{Alpha: 0.98}
	pids := [3]PID{
		{Kp: 4.5, Ki: 0.04, Kd: 0.40, Limit: 400}, // roll
		{Kp: 4.5, Ki: 0.04, Kd: 0.40, Limit: 400}, // pitch
		{Kp: 6.0, Ki: 0.10, Kd: 0.20, Limit: 400}, // yaw
	}

	dt := float64(loopDtMs) / 1000.0
	armed := false

	for {
		loopStart := time.Now()

		// Read sensors
		ax, ay, az, gx, gy, gz := readIMU()
		rc := readRC()

		// Arm / disarm logic
		if rc.Armed && !armed {
			armed = true
			for i := range pids {
				pids[i].Reset()
			}
		} else if !rc.Armed {
			armed = false
		}

		// Attitude estimate
		roll, pitch := filter.Update(ax, ay, az, gx, gy, dt)

		// PID control
		rollCmd  := pids[0].Compute(0.0, roll,  dt)
		pitchCmd := pids[1].Compute(0.0, pitch, dt)
		yawCmd   := pids[2].Compute(0.0, gz,    dt)

		// Motor mix
		var motorUS [4]uint32
		if armed {
			motorUS = mixQuadX(rc.Throttle, pitchCmd, rollCmd, yawCmd)
		} else {
			motorUS = [4]uint32{pwmMinUs, pwmMinUs, pwmMinUs, pwmMinUs}
		}

		// Write outputs
		for i, pin := range motorPins {
			setPWM(pin, motorUS[i])
		}

		// Rate limiting
		elapsed := time.Since(loopStart)
		target  := time.Duration(loopDtMs) * time.Millisecond
		if elapsed < target {
			time.Sleep(target - elapsed)
		}
	}
}
